### PR TITLE
Add from_int_btc method to Amount

### PR DESCRIPTION
**Pull request by yancyribbens**
Followup PR from https://github.com/rust-bitcoin/rust-bitcoin/pull/1811

Added a `const` associated function `from_int_btc()` for Amount.  `panic()` in const context is only available after 1.57+ so a work around is provided.

### CODE REVIEW

**Comment by yancyribbens** [comment]
hmm this compiles fine on my local machine but I didn't notice the warning "invalid feature rust_v_1_57 .." which causes CI to fail.  Looks like this needs more work.

**Comment by tcharding** [comment]
Should be using conditional configuration option not a feature (that means `#[cfg(rust_v_X_Y`). There is another instance of using it as a feature in the code currently which is wrong.

**Comment by apoelstra** [comment]
> or panic in debug mode happen

Well, we can't panic under any circumstances because we're trying to make a `constfn`.

We could overflow in both release and debug mode, I suppose.

**Comment by yancyribbens** [comment]
I'm afk right now so I can't test, however I was thinking const fn will still panic in debug mode if for example `u64::max * 2`  is calculated, otherwise in release mode it silently overflows.

**Comment by yancyribbens** [comment]
I tested on the rust playground and I guess since it's const fn the compiler won't allow overflows to compile.  Although I'm not sure how the sementics work with const fn if user args are provided.  Maybe that's not possible.

**Comment by tcharding** [comment]
How about this, although it leaks the internal representation in the docs, is that acceptable?
```rust
    /// Creates an [`Amount`] from a value representing a whole number of bitcoin.
    ///
    /// Internally computes the number of satoshis (ie, `btc * 100_000_000`)
    /// saturating at the numeric bounds of `u64`.
    pub const fn saturating_from_int_btc(btc: u64) -> Amount {
        Amount(btc.saturating_mul(100_000_000))
    }
```
and a unit test
```rust
    #[test]
    fn const_saturating_from_int_btc_overflow() {
        let got = Amount::saturating_from_int_btc(u64::MAX - 10);
        let want = Amount(u64::MAX);
        assert_eq!(got, want);
    }
```

**Comment by tcharding** [comment]
OT: Why don't we have a const `SATS_PER_BITCOIN` instead of all the `100_000_000` in the code base?

**Comment by junderw** [comment]
> OT: Why don't we have a const SATS_PER_BITCOIN instead of all the 100_000_000 in the code base?

`Amount::ONE_BTC` exists... maybe use that everywhere?

**Comment by junderw** [comment]
> How about this

If we make a saturating one, I'd expect it to saturate to a whole BTC... maybe even 21 mil (MAX_BTC)

u64::MAX is kind of weird, since it's not a whole bitcoin.

**Comment by tcharding** [comment]
> > How about this
> 
> If we make a saturating one, I'd expect it to saturate to a whole BTC... maybe even 21 mil (MAX_BTC)
> 
> u64::MAX is kind of weird, since it's not a whole bitcoin.

I was under the impression that everyone knows that bitcoin (in general not just in rust-bitcoin) uses ints not floats so this behaviour is obvious but that's coming from the perspective of a low level C guy with a history of disliking floats - am I wrong?

**Comment by junderw** [comment]
I just think it's weird that

`184467440737` will result in `Amount(18446744073700000000)`
but
`184467440738` and above will result in `Amount(18446744073709551615)`

Since the whole point of the function is to turn `10` into `10 BTC worth of an Amount`...

But again, I think this is just personal preference, and if it's documented anything is fine.

**Comment by tcharding** [comment]
Cool, thanks for getting the bounds. So if I understand you clearly this leads to something like (excluding the exact name):
```rust
pub const: u64 MAX_EXPRESSIBLE_WHOLE_BITCOIN_IN_SATS = 18446744073700000000;
```

and we saturate to that. I agree with the sentiment, the reason I'd like to avoid this is because it provides an answer to the "what is the meaning of max bitcoin" question. Currently we implicitly define max bitcoin as `u64::MAX` sats and so far we have been unable to get any improvement on that, at least that is how I understood the debate.

What if, in this PR, we saturate to `Amount::MAX`, state as such in the docs, then leave the whole "what is Amount::MAX, and what does it even mean" discussion for later work on the `amount` module (which is getting a bunch of work at the moment). Would that be a good compromise?

**Comment by junderw** [comment]
Sounds perfect.

BTC_MAX (21 mil BTC) and MAX seems confusing though.

**Comment by apoelstra** [comment]
concept ACK from me on saturating as long as it's documented. The warning should probably be in bold, and we should briefly explain that we'd like to just panic but can't in a constfn at our current MSRV.

I also think that we should explore whether we can just use `*` rather than `saturating_mul`, and if that will give us the desired "panic in debug mode, overflow in release mode, refuse to compile in a compile-time-evaluated constfn" behavior.

**Comment by tcharding** [comment]
You continue prodding finally tweaked my interest, it behaves as we want - see: #1883

**Comment by Kixunil** [comment]
I'm not convinced saturating is that useful, we should just use the array out of bounds trick to panic in const.

**Comment by yancyribbens** [comment]
I also don't see why we should bother with saurating_mul.  As I understand, const context evaluates when compiled, and if there is an overflow it's better to have a compiler error and not compile instead of saturating.

**Comment by yancyribbens** [comment]
> I'm not convinced saturating is that useful, we should just use the array out of bounds trick to panic in const.

It looks like there is a lint that prevents the array oob trick :facepalm:

**Comment by yancyribbens** [comment]
If the OOB trick isn't possible, maybe feature gate this so it's only available in 1.57+?

**Comment by apoelstra** [comment]
You could either disable the lint or just remove the `let` binding.

**Comment by Kixunil** [comment]
Just disable the lint with a comment. Removing `let` wouldn't work and would trigger another lint.

**Comment by yancyribbens** [comment]
Sorry for the delay, was afk for a bit

**Review by tcharding — COMMENTED** [review]
Thanks man, looks good!

ACK  b7b5e0452f337c442640bb870488b666059b32d3

There are a bunch more places this new function could be used if you get super motivated (`git grep _000_000`).

**Review by tcharding — APPROVED** [review]

File: bitcoin/Cargo.toml [hunk 1]
```diff
@@ -21,6 +21,7 @@ secp-lowmemory = ["secp256k1/lowmemory"]
 secp-recovery = ["secp256k1/recovery"]
 bitcoinconsensus-std = ["bitcoinconsensus/std", "std"]
 
+
```
Comment (tcharding) [inline bitcoin/Cargo.toml:24]: This shouldn't really be in here but I won't hold you up for it :)

Comment (yancyribbens) [reply bitcoin/Cargo.toml:24]: Doh, will fix

**Comment by yancyribbens** [comment]
> git grep _000_000

Oh nice, git grep is dope.  Or better yet, `git grep -n _000_000` so one can find the line num.

**Comment by tcharding** [comment]
I use a shell alias to `git grep -IPn --color=always`

**Review by tcharding — APPROVED** [review]
ACK 88df0115f5b919195fd6dced10f56902ee380bb9

File: bitcoin/src/amount.rs [hunk 2]
```diff
@@ -517,6 +517,22 @@ impl Amount {
     #[deprecated(since = "0.31.0", note = "Use Self::MIN instead")]
     pub const fn min_value() -> Amount { Amount(u64::min_value()) }
 
+    /// Convert from a value expressing integer values of bitcoins to an [Amount] in const context.
```
Comment (Kixunil) [inline bitcoin/src/amount.rs:525]: Should have `# Panics` section.

Comment (tcharding) [reply bitcoin/src/amount.rs:525]: Lucky we have @Kixunil on review to catch the things that I miss :)

**Comment by yancyribbens** [comment]
> There are a bunch more places this new function could be used if you get super motivated (`git grep _000_000`).

This is the `const` version, so it's really only useful for other `const` functions.

File: bitcoin/src/amount.rs (hunk 2 above)
Comment (yancyribbens) [reply bitcoin/src/amount.rs]: Thanks, I added a panic section

**Comment by tcharding** [comment]
Woops :)

File: bitcoin/src/amount.rs [hunk 3]
```diff
@@ -201,7 +201,13 @@ impl std::error::Error for ParseAmountError {
...
 
 fn is_too_precise(s: &str, precision: usize) -> bool {
-    s.contains('.') || precision >= s.len() || s.chars().rev().take(precision).any(|d| d != '0')
+    match s.find('.') {
+        Some(pos) =>
+            s[(pos + 1)..].chars().any(|d| d != '0')
+                || precision >= pos
+                || s[..pos].chars().rev().take(precision).any(|d| d != '0'),
+        None => precision >= s.len() || s.chars().rev().take(precision).any(|d| d != '0'),
+    }
```
Comment (tcharding) [inline bitcoin/src/amount.rs:210]: Did you mean to put this in here? Its unrelated to the PR.

File: bitcoin/src/amount.rs [hunk 4]
```diff
@@ -1582,6 +1611,18 @@ mod tests {
...
     }
 
+    #[test]
+    fn from_int_btc() {
+		let amt = Amount::from_int_btc(2);
+		assert_eq!(Amount::from_sat(200_000_000), amt);
+	}
+
+	#[should_panic]
+    #[test]
```
Comment (tcharding) [inline bitcoin/src/amount.rs:1621]:
```suggestion
    #[should_panic]
    #[test]
```

File: bitcoin/src/amount.rs [hunk 5]
```diff
@@ -1582,6 +1611,18 @@ mod tests {
...
+    fn from_int_btc_panic() {
```
Comment (tcharding) [inline bitcoin/src/amount.rs:1622]:
```suggestion
    fn from_int_btc_overflow() {
```

**Review by tcharding — COMMENTED** [review]
A bunch of unrelated changes have snuck into the PR, can you remove them or explain them please. (The PR describtion is stale now too.)

Thanks man

File: bitcoin/src/amount.rs (hunk 3 above)
Comment (yancyribbens) [reply bitcoin/src/amount.rs:210]: Crap, looks like a bad git rebase.

**Comment by yancyribbens** [comment]
> A bunch of unrelated changes have snuck into the PR, can you remove them or explain them please? (The PR describtion is stale now too.)
> 
> Thanks man

Sorry about that.  It looks like I pulled in some old state from master on accident.

**Review by apoelstra — APPROVED** [review]
ACK 9f7449b57299e550bae1107f96d21a95c7c684af

**Pull request merged**